libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -105,7 +106,10 @@ bool StaticAnalyzer::visit(Return const& _return)
 bool StaticAnalyzer::visit(ExpressionStatement const& _statement)
 {
 	if (_statement.expression().annotation().isPure)
-		warning(_statement.location(), "Statement has no effect.");
+		m_errorReporter.warning(_statement.location(),
```

Comment:
Oh and if you want to restrict e.g. the TypeChecker to only report typeErrors, you can add another class in between that provides the `typeError(...)` function that internally calls `error(TypeError, ...)`